auth/src/main/java/software/amazon/awssdk/auth/signer/Aws4Signer.java

```diff
@@ -15,478 +15,99 @@
-    }
-
-    private String getCanonicalizedHeaderString(Map<String, List<String>> headers) {
-        final List<String> sortedHeaders = new ArrayList<>(headers.keySet());
-        sortedHeaders.sort(String.CASE_INSENSITIVE_ORDER);
+    public SdkHttpFullRequest presign(SdkHttpFullRequest requestToSign, ExecutionAttributes executionAttributes) {
+        Aws4PresignerParams signingParams = constructAwsPresignerParams(executionAttributes);
 
-        final Map<String, List<String>> requestHeaders = headers;
-        StringBuilder buffer = new StringBuilder();
-        for (String header : sortedHeaders) {
-            if (shouldExcludeHeaderFromSigning(header)) {
-                continue;
-            }
-            String key = lowerCase(header);
-
-            for (String headerValue : requestHeaders.get(header)) {
-                StringUtils.appendCompactedString(buffer, key);
-                buffer.append(":");
-                if (headerValue != null) {
-                    StringUtils.appendCompactedString(buffer, headerValue);
-                }
-                buffer.append("\n");
-            }
-        }
-
-        return buffer.toString();
+        return presign(requestToSign, signingParams);
     }
 
-    private String getSignedHeadersString(Map<String, List<String>> headers) {
-        final List<String> sortedHeaders = new ArrayList<>(headers.keySet());
-        sortedHeaders.sort(String.CASE_INSENSITIVE_ORDER);
-
-        StringBuilder buffer = new StringBuilder();
-        for (String header : sortedHeaders) {
-            if (shouldExcludeHeaderFromSigning(header)) {
-                continue;
-            }
-            if (buffer.length() > 0) {
-                buffer.append(";");
-            }
-            buffer.append(lowerCase(header));
+    public SdkHttpFullRequest presign(SdkHttpFullRequest request, Aws4PresignerParams signingParams) {
+        // anonymous credentials, don't sign
+        if (CredentialUtils.isAnonymous(signingParams.awsCredentials())) {
+            return request;
         }
 
-        return buffer.toString();
-    }
+        final Aws4SignerRequestParams requestParams = new Aws4SignerRequestParams(signingParams);
 
-    private boolean shouldExcludeHeaderFromSigning(String header) {
-        return LIST_OF_HEADERS_TO_IGNORE_IN_LOWER_CASE.contains(lowerCase(header));
+        return doPresign(request, requestParams, signingParams).build();
     }
 
-    private void addHostHeader(SdkHttpFullRequest.Builder mutableRequest) {
-        // AWS4 requires that we sign the Host header so we
-        // have to have it in the request by the time we sign.
+    private Aws4PresignerParams constructAwsPresignerParams(ExecutionAttributes executionAttributes) {
+        final Aws4PresignerParams.Builder presignerParams =
+            Aws4PresignerParams.builder()
+                               .awsCredentials(executionAttributes.getAttribute(AwsExecutionAttributes.AWS_CREDENTIALS))
```

Comment:
Made a similiar comment on a previous revision. You can have PresignerParams just contain the Aws4SignerParams so you don't have to duplicate the extraction here.